dkg/pedersen/handler.go

```diff
@@ -513,5 +656,321 @@ func (h *Handler) handleDeal(ctx context.Context, msg types.Deal, out mino.Sende
+				deal.Deal.DHKey,
+				deal.Deal.Signature,
+				deal.Deal.Nonce,
+				deal.Deal.Cipher,
+			),
+		)
+
+		//dealResharing contains the public coefficients as well
+		dealResharingMsg := types.NewReshare(dealMsg, publicCoeff)
+
+		h.log.Trace().Msgf("%s sent dealResharing %d", h.me, i)
+
+		select {
+		case err := <-out.Send(dealResharingMsg, participants[i]):
+			if err != nil {
+				return xerrors.Errorf("failed to send resharing deal: %v", err)
+			}
+		case <-ctx.Done():
+			return xerrors.Errorf("context done: %v", ctx.Err())
+		}
+	}
+
+	h.log.Debug().Msgf("%s sent all its deals", h.me)
+
+	return nil
+}
+
+// receiveDealsResharing is similar to receiveDeals except that it receives the
+// dealResharing. Only the new or common nodes call this function
+func (h *Handler) receiveDealsResharing(ctx context.Context, isCommonNode bool, resharingRequest types.ResharingRequest, out mino.Sender, reshares cryChan[types.Reshare]) error {
+	h.log.Trace().Msgf("%v is handling deals from other nodes", h.me)
+
+	addrsNew := resharingRequest.GetAddrsNew()
+	var addrsOld []mino.Address
+
+	numReceivedDeals := 0
+	expectedDeals := 0
+
+	// by default we read the old addresses from the state
+	addrsOld = h.startRes.GetParticipants()
+
+	// the new nodes don't have the old addresses in their state
+	if !isCommonNode {
+		addrsOld = resharingRequest.GetAddrsOld()
+	}
+
+	expectedDeals = len(addrsOld)
+
+	// we find the union of the old and new address sets to aviod from sending a
+	// message to the common nodes multiple times
+	addrsAll := unionOfTwoSlices(addrsNew, addrsOld)
+
+	for numReceivedDeals < expectedDeals {
+		reshare, err := reshares.pop(ctx)
+		if err != nil {
+			return xerrors.Errorf("context done: %v", err)
+		}
+
+		if !isCommonNode && numReceivedDeals == 0 {
+
+			c := &pedersen.Config{
+				Suite:        suite,
+				Longterm:     h.privKey,
+				OldNodes:     resharingRequest.GetPubkeysOld(),
+				NewNodes:     resharingRequest.GetPubkeysNew(),
+				PublicCoeffs: reshare.GetPublicCoeffs(),
+				Threshold:    resharingRequest.GetTNew(),
+				OldThreshold: resharingRequest.GetTOld(),
+			}
+
+			newDkg, err := pedersen.NewDistKeyHandler(c)
+			if err != nil {
+				return xerrors.Errorf("failed to create the new dkg for %s: %v",
+					h.me, err)
+			}
+
+			h.dkg = newDkg
+		}
+
+		deal := reshare.GetDeal()
+
+		err = h.handleDeal(ctx, deal, out, addrsAll)
+		if err != nil {
+			h.log.Warn().Msgf("%s failed to handle deal: %v", h.me, err)
+			return xerrors.Errorf("failed to handle received deal: %v", err)
+		}
+
+		numReceivedDeals++
+	}
+
+	h.log.Debug().Msgf("%v received all the expected deals", h.me)
+
+	return nil
+}
+
+///////// auxiliary functions
+
+// isInSlice gets an address and a slice of addresses and returns true if that
+// address is in the slice. This function is called for cheking whether an old
+// committee member is in the new commmitte as well or not
+func isInSlice(addr mino.Address, addrs []mino.Address) bool {
```

Comment:
If we move this to a separate package, I'd adopt the more typical API `func Contains[E comparable](s []E, v E) bool`.
As a matter of fact, that's what might get integrated in the standard library at some point
https://pkg.go.dev/golang.org/x/exp/slices#Contains